presto-main/src/main/java/com/facebook/presto/operator/project/DictionaryAwarePageProjection.java

```diff
@@ -0,0 +1,144 @@
+        verify(projection.getInputChannels().size() == 1, "projection must have only one input");
+    }
+
+    @Override
+    public Type getType()
+    {
+        return projection.getType();
+    }
+
+    @Override
+    public boolean isDeterministic()
+    {
+        return projection.isDeterministic();
+    }
+
+    @Override
+    public InputChannels getInputChannels()
+    {
+        return projection.getInputChannels();
+    }
+
+    @Override
+    public Block project(ConnectorSession session, Page page, SelectedPositions selectedPositions)
+    {
+        Block block = page.getBlock(0);
+        if (block instanceof LazyBlock) {
+            block = ((LazyBlock) block).getBlock();
+        }
+
+        if (block instanceof RunLengthEncodedBlock) {
+            Block value = ((RunLengthEncodedBlock) block).getValue();
+            Optional<Block> projectedValue = processDictionary(session, value);
+            // single value block is always considered effective
+            verify(projectedValue.isPresent());
+            return new RunLengthEncodedBlock(projectedValue.get(), selectedPositions.size());
+        }
+
+        if (block instanceof DictionaryBlock) {
+            DictionaryBlock dictionaryBlock = (DictionaryBlock) block;
+            // Attempt to process the dictionary.  If dictionary is processing has not been considered effective, an empty response will be returned
+            Optional<Block> projectedDictionary = processDictionary(session, dictionaryBlock.getDictionary());
+            // record the usage count regardless of dictionary processing choice, so we have stats for next time
+            lastDictionaryUsageCount += selectedPositions.size();
+            // if dictionary was processed, produce a dictionary block; otherwise do normal processing
+            if (projectedDictionary.isPresent()) {
+                lastDictionaryUsageCount += selectedPositions.size();
+                int[] outputIds = filterDictionaryIds(dictionaryBlock, selectedPositions);
+                return new DictionaryBlock(selectedPositions.size(), projectedDictionary.get(), outputIds, false, sourceIdFunction.apply(dictionaryBlock));
+            }
+        }
+
+        return projection.project(session, new Page(block), selectedPositions);
+    }
+
+    private Optional<Block> processDictionary(ConnectorSession session, Block dictionary)
+    {
+        if (lastInputDictionary == dictionary) {
+            return lastOutputDictionary;
+        }
+
+        // Process dictionary if:
+        //   this is the first block
+        //   there is only entry in the dictionary
+        //   the last dictionary was used for more positions than were in the dictionary
+        boolean shouldProcessDictionary = lastInputDictionary == null || dictionary.getPositionCount() == 1 || lastDictionaryUsageCount >= lastInputDictionary.getPositionCount();
+
+        lastDictionaryUsageCount = 0;
```

Comment:
I am curious what led you to this shape of algorithm for turning on and off dictionary processing. It does not seem natural and is hard to follow for me.
I would probably go for sth based on exponential regression. In the spirit of what you have for determining number of positions to process in PageProcessor. 

Some explanation in javadoc of this class or (new super class) would be helpful.